core/src/main/java/org/everit/json/schema/ArraySchema.java

```diff
@@ -204,57 +206,88 @@ public boolean requiresArray() {
             if (schemaOfAdditionalItems != null) {
-                validateItemsAgainstSchema(IntStream.range(itemValidationUntil, subject.length()),
-                        subject,
-                        schemaOfAdditionalItems,
-                        rval::add);
+                validateItemsAgainstSchema(itemValidationUntil, subject.length(), subject, schemaOfAdditionalItems,
+                        new RValAddConsumer(rval));
             }
         }
         return rval;
     }
 
-    private void validateItemsAgainstSchema(final IntStream indices, final JSONArray items,
-            final Schema schema,
+    private void validateItemsAgainstSchema(final int endExclusive, final JSONArray items, final Schema schema,
             final Consumer<ValidationException> failureCollector) {
-        validateItemsAgainstSchema(indices, items, i -> schema, failureCollector);
+        validateItemsAgainstSchema(0, endExclusive, items, schema, failureCollector);
     }
 
-    private void validateItemsAgainstSchema(final IntStream indices, final JSONArray items,
-            final IntFunction<Schema> schemaForIndex,
+    private void validateItemsAgainstSchema(final int startInclusive, final int endExclusive, final JSONArray items,
+            final Schema schema,
             final Consumer<ValidationException> failureCollector) {
-        for (int i : indices.toArray()) {
-            String copyOfI = String.valueOf(i); // i is not effectively final so we copy it
+        validateItemsAgainstSchema(startInclusive, endExclusive, items, new Function<Integer, Schema>() {
+            @Override
+            public Schema apply(Integer input) {
+                return schema;
+            }
+        }, failureCollector);
+    }
+
+    private void validateItemsAgainstSchema(final int endExclusive, final JSONArray items,
+            final Function<Integer, Schema> schemaForIndex, final Consumer<ValidationException> failureCollector) {
+        validateItemsAgainstSchema(0, endExclusive, items, schemaForIndex, failureCollector);
+    }
+
+    private void validateItemsAgainstSchema(final int startInclusive, final int endExclusive, final JSONArray items,
+            final Function<Integer, Schema> schemaForIndex, final Consumer<ValidationException> failureCollector) {
+        for (int i = startInclusive; i < endExclusive; i++) {
+            final String copyOfI = String.valueOf(i); // i is not effectively final so we copy it
             ifFails(schemaForIndex.apply(i), items.get(i))
-                    .map(exc -> exc.prepend(copyOfI))
-                    .ifPresent(failureCollector);
+                    .transform(new Function<ValidationException, ValidationException>() {
+                        @Override
+                        public ValidationException apply(ValidationException exc) {
+                            ValidationException exception = exc.prepend(copyOfI);
+                            failureCollector.accept(exception);
```

Comment:
the whole `transform` block of `Optional` runs only if the exception isn't null. I know it's counter-intuitive, if you want I can move it.